### DISCUSSION TITLE
Audio stuttering over serial

### DISCUSSION THREAD

**Discussion by randomdude583** [pschatzmann/arduino-audio-tools#1805] [Q&A] [answered]
I am brand new to arduino-audio-tools, and pretty new to microcontrollers in general. I am currently trying to generate a sine wave in python and pass it over serial to the arduino, then play it on a headphone. I believe I should have that working, but it is extremely choppy, cutting in and out multiple times a second.

As a test, I have a modified version of the send-8bit-receive.ino example that outputs to a headphone and uses UART0, which is the one exposed over USB. It works as expected when bridging TX and RX. However, if I use a very simple python script to echo serial back out, it is still super choppy.
I have messed around with setting buffer size on the streamCopy, but it hasn't had a noticeable effect. 

Any idea what could be causing this?

**Comment by pschatzmann** [comment, marked as answer]
Try to measure the thruput that you can achieve via your connection.
From that you can calculate the max sample rate!

It can also be that your host is causing the choppiness, then you might need to add a buffer on the receiving size and fill it before you start.

**Reply by randomdude583** [reply]
I created a test to be sure that the two can communicate at 300000 baud, and they do in fact have an issue, I think due to clock inaccuracy. I tested them to be stable up to 921600, so I switched over to a more standard baudrate of 460800, which I also verified to be working. Unfortunately the audio is still extremely choppy at this baudrate. 

As for the buffer, are you referring to the buffer size in the streamCopy, or am I missing something?

Thanks!

**Comment by pschatzmann** [comment]
No from the readme: Different [Buffer Implementations](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_base_buffer.html) for [QueueStream](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_queue_stream.html)

**Reply by randomdude583** [reply]
I'll check that out. Do you know off the top of your head if there is an example that implements this?
